src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs

```diff
@@ -890,5 +891,655 @@ static void Main(string[] args)
+    public class {|#0:Foo|} : IFoo
+    {
+        [RequiresPreviewFeatures]
+        decimal IFoo.Value => 1.1m; // No diagnostic
+    }
+
+    [RequiresPreviewFeatures]
+    interface IFoo
+    {
+        decimal Value { get; }
+    }
+}";
+
+            var test = TestCS(csInput);
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(0).WithArguments("Foo"));
+            await test.RunAsync();
+
+        }
+
+        [Fact]
+        public async Task TestNestedClass()
+        {
+            var csInput = @" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{" +
+@"
+
+    class Program
+    {
+        [RequiresPreviewFeatures]
+        class NestedClass
+        {
+
+        }
+
+        static void Main(string[] args)
+        {
+            NestedClass nestedClass = {|#0:new NestedClass()|};
+        }
+    }
+}";
+
+            var test = TestCS(csInput);
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(0).WithArguments(".ctor"));
+            await test.RunAsync();
+
+        }
+
+        [Fact]
+        public async Task TestGenericMethodWithPreviewClass()
+        {
+            var csInput = @" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{" +
+@"
+
+    class Program
+    {
+        public bool {|#0:GenericMethod|}<T>()
+            where T : Foo
+        {
+            return true;
+        }
+
+        static void Main(string[] args)
+        {
+        }
+    }
+
+    [RequiresPreviewFeatures]
+    public class Foo
+    {
+    }
+
+}";
+
+            var test = TestCS(csInput);
+            test.ExpectedDiagnostics.Add(VerifyCS.Diagnostic(DetectPreviewFeatureAnalyzer.GeneralPreviewFeatureAttributeRule).WithLocation(0).WithArguments("GenericMethod"));
+            await test.RunAsync();
+        }
+
+        [Fact]
+        public async Task TestGenericMethodInsidePreviewClass()
+        {
+            var csInput = @" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{" +
+@"
+
+    [RequiresPreviewFeatures]
+    class Program
+    {
+        public bool GenericMethod<T>()
+            where T : Foo
+        {
+            return true;
+        }
+
+        static void Main(string[] args)
+        {
+        }
+    }
+
+    [RequiresPreviewFeatures]
+    public class Foo
+    {
+    }
+
+}";
+
+            var test = TestCS(csInput);
+            await test.RunAsync();
+
+        }
+
+        [Fact]
+        public async Task TestTwoLevelGenericMethodInsidePreviewClass()
+        {
+            var csInput = @" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{" +
+@"
+
+    [RequiresPreviewFeatures]
+    class Program
+    {
+        class NestedClass
+        {
+            public bool {|#1:GenericMethod|}<T>()
```

Comment:
Isn't it false positive

Reply:
I don't think so. We only 1 level look up for methods/properties, so any other level of nesting will produce the diagnostic. I also didn't spend too much time trying to optimize this since either we expect users to just opt into preview features at the assembly level.

Reply:
For me this is a false positive, the parent class annotation should suppress this. Though I am not sure if suppressing could only one level up is by design cc @terrajobst @jeffhandley

Reply:
I do think we should look "all the way up" the hierarchy. If an API is contained in a preview feature scope, it too should be recognized as a preview feature, regardless of how deep into the hierarchy it is.

Reply:
Follow-up... we should look all the way up to the assembly even, such that if an assembly is marked, everything in the assembly can call preview features.

Reply:
Fixed now.